### PR TITLE
[Exam] File upload exercises for exam mode

### PR DESCRIPTION
**Pull request by kloessst**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I added multiple integration tests (Spring) related to the features
- [x] Server: I implemented the changes with a good performance and prevented too many database calls
- [x] Server: I documented the Java code using JavaDoc style.
- [x] Client: I added multiple integration tests (Jest) related to the features
- [x] Client: I added `authorities` to all new routes and check the course groups for displaying navigation elements (links, buttons)
- [x] Client: I documented the TypeScript code using JSDoc style.
- [x] Client: I added multiple screenshots/screencasts of my UI changes
- [x] Client: I translated all the newly inserted strings into German and English

### Motivation and Context
File upload exercise CRUD operations must be adapted for exam mode.

### Description
Client:
- [x] Created new client routes for text-exercise update and creation
For creation: ':examId/exercise-groups/:groupId/file-upload-exercises/new'
For update: ':examId/exercise-groups/:groupId/file-upload-exercises/:exerciseId/edit'
- [x] Load file upload exercise module
- [x] Extend FileUploadExerciseResolver by fetching of exerciseGroup for the new routes
- [x] Deactivate categories in exam mode, as this relies on a Course
- [x] Added "Add file upload exercise" button to exercise group component 
- [x] Refactored the TextExerciseRowButtonsComponent. This component is responsible for displaying  edit and delete button of exam exercises. The component can now be used for all exercise types.
- [x] Added exam mode to TextExercise and FileUploadExercise Detail Component. 

Server:
- [x] Check that course and exerciseGroup is never set simultaneously for creation/update
- [x] Extend exercise creation/update authorization check for exam mode
- [x] Deactivate Tutor/Student notification on exercise creation/update
- [x] Add additional authorization check to GET file upload exercise route. Only instructors and admins should be allowed to access an exam file upload exercise
- [x] GET for exam text and file upload exercises now also returns the exam and course. The client can use this to build the right routes for example for the Edit button of the Detail Component
- [x] Adapted authorization check for deletion
- [x] Refactored course retrieval over ExerciseGroups

**We still have a problem with client route consistency. For example we only use the exerciseId of edit routes such as `course-management/:courseId/exams/:examId/exercise-groups/:groupId/file-upload-exercises/:exerciseId/edit` to retrieve the exercise. The other id are not used and could be anything. This is not a security problem as the server still does all the authorization checks. Nevertheless this should be fixed in the future**

### Steps for Testing
_Ensure that the existing CRUD operations for regular file upload exercises still work_

1. Log in as admin or instructor
2. Create a regular file upload exercise
--> Exercise was created
3. Update the exercise
--> Exercise was updated
4. View the exercise
--> Can open the page
5. Delete the exercise
--> Exercise was deleted

_Ensure that CRUD operations for exam file upload exercises work_

1. Log in as admin or instructor
2. Create an exam
3. Create an exercise group
4. Create a file upload exercise
--> Exercise was created
5. Edit the file upload exercise
--> Can update the file upload exercise
6. Navigate to `/course-management/:courseId/file-upload-exercises/:exerciseId` using the id of an exam file upload exercise
--> Should be able to see the exercise
7. Delete the exercise
--> Exercise was deleted

_Ensure correct authorization for file upload exercises_
1. Log in as a tutor or student
2. Navigate to `course-management/:courseId/exams/:examId/exercise-groups/:groupId/file-upload-exercises/new`
--> You should not be able to access the route
4. Use the id of an exam exercise for the following steps
5. Navigate to `course-management/:courseId/exams/:examId/exercise-groups/:groupId/file-upload-exercises/:exerciseId/edit`
--> You should not be able to access the route
7. Navigate to `/course-management/:courseId/file-upload-exercises/:exerciseId`
--> You should not be able to see the exercise

Please also test regular and exam text exercises again as the component which displays buttons such as edit, view, teams (for regular exercise) was refactored.

### Screenshots
fileupload

### CODE REVIEW

**Comment by pzdr7** [comment]
Navigating to /course-management/:courseId/file-upload-exercises/:exerciseId for an exam file upload exercise and hitting the edit button there does nothing & causes an error for me: 

image

Is there anything we can do about that?

**Comment by kloessst** [comment]
@pzdr7 Currently working on a solution. The problem here is that the edit button has a hard coded route to the edit for normal exercises at the moment.

**Comment by pzdr7** [comment]
I have tested this on the test server 1 (more specifically, I created an exam, an exercise group and a file upload and text exercise, viewed them using the route you specified and then tried to access the same routes using a student and a tutor account) and I haven't found any new issues apart from the course missing from the view at /course-management/:courseId/file-upload-exercises/:exerciseId (see my previous comment's screenshot). This also occurs for text exercises.

**Comment by krusche** [comment]
> I have tested this on the test server 1 (more specifically, I created an exam, an exercise group and a file upload and text exercise, edited&viewed them using the route you specified and then tried to access the same routes using a student and a tutor account) and I haven't found any new issues apart from the course missing from the view at /course-management/:courseId/file-upload-exercises/:exerciseId (see my previous comment's screenshot). This also occurs for text exercises.
> 
> As for deletion of exercises: This still appears to be functional.

How did you navigate here? By manually inserting the URL?
It should not be possible to navigate to this page normally on Artemis.

Have you tested if you can still create file upload exercises normally (within a course)?

**Comment by pzdr7** [comment]
> How did you navigate here? By manually inserting the URL?
> It should not be possible to navigate to this page normally on Artemis.

I used the URL to navigate there, so I suppose it's not too much of a problem (this page is concealed, anyway).

> Have you tested if you can still create file upload exercises normally (within a course)?

Yes, and I've also tested if I can edit and delete them.

**Comment by kloessst** [comment]
@pzdr7 @krusche  At the moment the "view" components do not know, that the course is hidden inside exerciseGroup -> exam -> course. We can change this later on for all exercise types in one go.

File: src/main/webapp/app/course/manage/course-management.component.ts [hunk 1]
```diff
@@ -21,7 +21,7 @@ export class CourseManagementComponent implements OnInit, OnDestroy, AfterViewIn
     predicate: string;
     reverse: boolean;
     showOnlyActive = true;
-    showExamButton = false;
+    showExamButton = true;
```
Comment (sascha11110) [inline src/main/webapp/app/course/manage/course-management.component.ts:24]: Should be reverted on release :-)

**Pull request merged**